### ISSUE TITLE
[move] Add choice of anchor as well as more predefined positions

### ISSUE THREAD

**Issue by suuuehgi** [i3/i3#3689]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

* `move` command is relative to the upper left corner.
* `move position center` is the only fixed / predefined position (and relative to the upper left corner).

## Desired Behavior

* Choose anchor point (point of reference)
E.g.: `move <left|right|down|up> [<px> px] [anchor <tl|tr|bl|br|c>]`
* Multiple possible positions (that imply different anchor points)
E.g.: `move position <bottom|center|top> <left|center|right>`

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 4.16.1
</pre>

<pre>
- Linux Distribution & Version: Debian testing (buster), 4.19.0-4-amd64
- Are you using a compositor (e.g., xcompmgr or compton): No
</pre>

Thanks a lot!

**Comment by Airblader** [comment]
The "Current Behavior" description sounds like you are talking exclusively about moving *floating* windows. Is this correct?

> E.g.: move <left|right|down|up> [<px> px] [anchor <tl|tr|bl|br|c>]

Could you explain what all of these would do?

> E.g.: move position <bottom|center|top> <left|center|right>

All of these can be trivially achieved using external scripts. The "move center" one is the odd one out because i3 itself does actually need this (floating, scratchpad).

**Comment by suuuehgi** [comment]
> The "Current Behavior" description sounds like you are talking exclusively about moving floating windows. Is this correct?

Yes, sorry I forgot that.

>Could you explain what all of these would do?
>
>>E.g.: `move position <bottom|center|top> <left|center|right>`

Move a floating window to a position from `{bottom, center, top} X {left, center, right}`.

**E.g.** `move position center right`: The center of the right side of my floating window attached / aligned with the center of the right side of my screen.

![draft](https://user-images.githubusercontent.com/7861774/56475446-a51d4700-6488-11e9-81f3-0e1397026460.png)

It would be very handy for me not to rely on extra scripts and extra programs doing the calculations since this is window size and screen resolution dependent.

**Comment by Airblader** [comment]
> It would be very handy for me not to rely on extra scripts and extra programs doing the calculations since this is window size and screen resolution dependent.

But scripts can easily account for that?

**Comment by suuuehgi** [comment]
> But scripts can easily account for that?

That depends on the definition of _easy_ or rather the viewpoint. Someone working on the code base could maybe do this "easily".

Afaik would I have to write a separate script for every application and after 10 min of googling I still don't know how to get the window dimensions for the calculations although I once wrote something similar (I can not find anymore).
So, from my pov, maybe simple but definitely not easy and definitely not comfortable.

**Comment by Airblader** [comment]
> Afaik would I have to write a separate script for every application

I wouldn't see why the application matters. The script just has to figure out the current window, its dimensions, the screen dimensions, do whatever math you want to do and reposition the window. All of this can be done using [i3's IPC](https://i3wm.org/docs/ipc.html), using a lot of open tools such as `xdotool`, or by speaking to X directly (using Xlib or xcb).

I'm not too inclined to add more commands to i3 here because we are not a floating window manager and try to limit floating window support to the necessary things. We already allow `move position` to exact pixel-based positioning, which is a compromise as it keeps API surface small but allows for maximum flexibility.

If you need help writing such a script, you can head over to /r/i3wm where people might well be able to help you out. Chances are someone already wrote something like this, even.

**Comment by 0xC0ncord** [comment]
Would i3's IPC be able to allow a script to handle an event on a window's resizing? I was hoping that an anchoring feature as described by this issue would be possible. I have an application (Microsoft Teams) that creates its own desktop notifications in its own window. If multiple notifications are received in the same time frame, the notifications get stacked on top of each other in the same window, and the window resizes. I would like to be able to have the notification window anchored to the bottom right, and then be able to automatically move the window's position if the window is resized.

**Comment by Airblader** [comment]
What does it mean for it to create them "in the window"? If the notifications aren't top level windows, or are override redirect windows (which notifications usually are), they're not managed by i3.

Regardless, listening to resize can even be done window manager agnostic.

**Comment by 0xC0ncord** [comment]
> What does it mean for it to create them "in the window"? If the notifications aren't top level windows, or are override redirect windows (which notifications usually are), they're not managed by i3.
> 
> Regardless, listening to resize can even be done window manager agnostic.

Microsoft Teams notifications are delivered as a popup in their own window. If multiple notifications are received, the popup window simply resizes and renders all notifications together. They're not desktop notifications (delivered over D-Bus) in other words.

My question is whether or not i3's IPC can be used to listen to an event like this. The notification window isn't raised as urgent, and the window doesn't change positions if it resizes itself. I could simply poll the window list in a script and adjust the window's position every tick, but that seems extremely inefficient.

**Comment by Airblader** [comment]
> They're not desktop notifications (delivered over D-Bus) in other words.

Yeah, but they're still most likely not managed windows. Can you move them using i3 commands as normal floating windows? You can also make sure using `xwininfo` and clicking on that window (it'll tell you whether it's override-redirect).

If they're not managed windows, i3 doesn't handle them at all anyway.

> My question is whether or not i3's IPC can be used to listen to an event like this.

I don't think we send events on resize currently (and that would be quite a lot of events). But you can do it with pure X tools. Perhaps it's even possibly entirely with xdotool and no more advanced script would be necessary. I'm not sure on that, though.

> I could simply poll the window list in a script and adjust the window's position every tick, but that seems extremely inefficient.

Indeed, but X is mostly event-based, so you'd rather listen for the events.

**Comment by 0xC0ncord** [comment]
> Can you move them using i3 commands as normal floating windows?

I can.

> But you can do it with pure X tools. Perhaps it's even possibly entirely with xdotool and no more advanced script would be necessary.

I will give it a shot.
